# <font color='red'><b>Bootstrap assignment</b> </font>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

#importing random module 
from random import *
from tqdm import tqdm

from sklearn.tree import DecisionTreeRegressor

In [2]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [3]:
number_samples = 30

In [4]:
x.shape

(506, 13)

In [5]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

* <font color='red'><b> Note - While selecting the random 60% datapoints from the whole data, make sure that the selected datapoints are all exclusive, repetition is not allowed. </b></font>

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

## <font color='red'><b>A few key points</b></font>

* Remember that the datapoints used for calculating MSE score contain some datapoints that were initially used while training the base learners (the 60% sampling). This makes these datapoints partially seen (i.e. the datapoints used for calculating the MSE score are a mixture of seen and unseen data).
Whereas, the datapoints used for calculating OOB score have only the unseen data. This makes these datapoints completely unseen and therefore appropriate for testing the model's performance on unseen data.

* Given the information above, if your logic is correct, the calculated MSE score should be less than the OOB score.

* The MSE score must lie between 0 and 10.
* The OOB score must lie between 10 and 35.

* The difference between the left nad right confidence-interval values must not be more than 10. Make sure this is true for both MSE and OOB confidence-interval values.

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/OfcFrUP.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [4]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    
    Selecting_rows = sample(list(range(0,x.shape[0])),303) #select 303 row indices from 506 indices at random without replacement
    Replacing_rows = choices(Selecting_rows,k=203)  #select 203 row indices out of 303 indices at random with repalcement
    number_of_columns = choice(list(range(3,14)))  #select the number of columns for each row sample
    Selecting_columns = sample(list(range(0,13)),number_of_columns) #select the column indices without replacement. The number of column indces chosen must be equal to the number of columns (min = 3)

    sample_data = input_data[Selecting_rows,:][:,Selecting_columns] #create a row sample using the selected row indices and colum indiced
    target_of_sample_data = y[Selecting_rows] #get the corresponding target data
    
    #Replicating Data 
    
    Replicated_sample_data = input_data[Replacing_rows,:][:,Selecting_columns] #replicate the row samples
    target_of_replicated_sample_data = target_data[Replacing_rows]
    
    #Concatenating data
    
    final_sample_data = np.vstack([sample_data,Replicated_sample_data]) #stack the row samples (picked with and without replacement)
    final_target_data = np.vstack([target_of_sample_data.reshape(-1,1),target_of_replicated_sample_data.reshape(-1,1)]) #stack the target data
    print(final_sample_data.shape,final_target_data.shape,len(Selecting_rows),len(Selecting_columns))
    return final_sample_data,final_target_data,Selecting_rows,Selecting_columns  #return the final data sample,final target data,the row index without replacement and the column indices
    
    
    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [5]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

(506, 9) (506, 1) 303 9


True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [6]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

def gen_data_30_base_models(number_samples):
    #generating a data of 506 rows for each of the 30 base learners
    for i in range(number_samples):
        print(i)
        a,b,c,d = generating_samples(x,y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)

In [7]:
gen_data_30_base_models(30)

0
(506, 11) (506, 1) 303 11
1
(506, 7) (506, 1) 303 7
2
(506, 12) (506, 1) 303 12
3
(506, 12) (506, 1) 303 12
4
(506, 10) (506, 1) 303 10
5
(506, 5) (506, 1) 303 5
6
(506, 10) (506, 1) 303 10
7
(506, 8) (506, 1) 303 8
8
(506, 13) (506, 1) 303 13
9
(506, 11) (506, 1) 303 11
10
(506, 11) (506, 1) 303 11
11
(506, 6) (506, 1) 303 6
12
(506, 8) (506, 1) 303 8
13
(506, 12) (506, 1) 303 12
14
(506, 5) (506, 1) 303 5
15
(506, 8) (506, 1) 303 8
16
(506, 12) (506, 1) 303 12
17
(506, 7) (506, 1) 303 7
18
(506, 12) (506, 1) 303 12
19
(506, 13) (506, 1) 303 13
20
(506, 7) (506, 1) 303 7
21
(506, 13) (506, 1) 303 13
22
(506, 5) (506, 1) 303 5
23
(506, 6) (506, 1) 303 6
24
(506, 8) (506, 1) 303 8
25
(506, 13) (506, 1) 303 13
26
(506, 8) (506, 1) 303 8
27
(506, 6) (506, 1) 303 6
28
(506, 5) (506, 1) 303 5
29
(506, 12) (506, 1) 303 12


<font color='cyan'> <b>Grader function - 2 </b></font>

In [8]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [9]:
def gen_base_models(number_sample):
    #Train each of the base learner using the data prepared and store all the base lerners in a list
    model_list = []
    for i in range(number_samples):
        input_data = list_input_data[i]
        target_data = list_output_data[i].flatten()

        dt = DecisionTreeRegressor(max_depth=None)
        dt.fit(input_data,target_data)

        model_list.append(dt)
    return model_list

In [10]:
model_list = gen_base_models(30)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [11]:
def predict_outcome(input_data):
    '''Function to prdict a bagged output given a datapoint'''
    final_output_list = []
    for row in input_data:
        
        i = 0
        output_list = []
        for model in model_list:
            
            data = row[list_selected_columns[i]]
            output_value = model.predict(data.reshape(1,-1))
            output_list.append(output_value)
            i+=1
        final_bagged_output = np.median(np.array(output_list))
        final_output_list.append(final_bagged_output)
    return final_output_list



In [12]:
bagged_list = predict_outcome(x)

In [13]:
def MSE_computation(actual,pred):
    mse = mean_squared_error(actual,pred)
    return mse

In [14]:
mse = MSE_computation(y,bagged_list)
print('The MSE on train data is ',mse)

The MSE on train data is  0.09695158102766804


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [15]:
def oob_input():
    #computing the bagged output on oob data and store the outcome in a list
    input_data_index = list(range(0,x.shape[0]))

    data_index = 0
    final_oob_list = []
    for row in x:
        pred_outcome_list  = []
        i = 0
        for model in model_list:
            model_trained_rows = (list_selected_row[i])
            if data_index not in model_trained_rows:
                data = row[list_selected_columns[i]]
                pred_value = model.predict(data.reshape(1,-1))
                pred_outcome_list.append(pred_value)
            i+=1
        print('number of models which gave prediction ',len(pred_outcome_list))
        final_pred_value = np.median(np.array(pred_outcome_list))
        final_oob_list.append(final_pred_value)

        data_index+=1
        
    return final_oob_list
    

In [ ]:
final_oob_list = oob_input()

In [17]:


#This is same as caluclationg  the MSE between actual data points and oob calculated
print('MSE on doing oob computation is ',MSE_computation(y,final_oob_list))

MSE on doing oob computation is  11.84766304347826


# <font color='blue'><b>Task 2</b></font>

In [ ]:
mse_list = []
oob_score_list = []
for cnt in tqdm(range(35)):
    list_input_data =[] #list to store overall input
    list_output_data =[] #list to store overall target
    list_selected_row= [] #list to store uniquely selected rows
    list_selected_columns=[] #list to store unique selected columns (min =3,max=13)
    
    gen_data_30_base_models(30)  #first generate row sampled and column sampled data for 30 base learners
    model_list = gen_base_models(30) #build 30 models each of them trained with respective data
    
    bagged_list = predict_outcome(x) #predict the value on each train datapoint
    mse = MSE_computation(y,bagged_list) #compute MSE
    mse_list.append(mse)
    
    final_oob_list = oob_input() #predict outcome on each OOB datapoint
    oob_score = MSE_computation(y,final_oob_list) #compute MSE
    oob_score_list.append(oob_score)
    
    

In [19]:
def check(iterations):
    assert(len(mse_list) == iterations and len(oob_score_list) == iterations)
    return True
check(35)

True

In [21]:
import math

mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

mean_oob_score = np.mean(oob_score_list)
std_oob_score = np.mean(oob_score_list)

lower_limit_mse = mean_mse-(2*(std_mse/(math.sqrt(len(mse_list)))))
upper_limit_mse = mean_mse+(2*(std_mse/(math.sqrt(len(mse_list)))))

lower_limit_oob = mean_oob_score-(2*(std_oob_score/(math.sqrt(len(oob_score_list)))))
upper_limit_oob = mean_oob_score+(2*(std_oob_score/(math.sqrt(len(oob_score_list)))))

print('The 95% CI of MSE across 35 samples is [',lower_limit_mse,',',upper_limit_mse,']')
print('The 95% CI of OOB score across 35 samples is [',lower_limit_oob,',',upper_limit_oob,']')

The 95% CI of MSE across 35 samples is [ 0.08057169593983873 , 0.13338769798516398 ]
The 95% CI of OOB score across 35 samples is [ 8.786385308361155 , 17.761059773656793 ]


In [27]:
def limit_check(ll,ul):
    assert((ul-ll) <= 10)
    return True

print('MSE check')
limit_check(lower_limit_mse,upper_limit_mse)
print('OOB check')
limit_check(lower_limit_oob,upper_limit_oob)

MSE check
OOB check


True

# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [24]:
#use the custom Bagging regressor on a given query point
xq= np.array([[0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]])
final_list = predict_outcome(xq)
print('The final predicted value from the Decision Tree Regressor is ',final_list[0])

The final predicted value from the Decision Tree Regressor is  19.1


<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

***Observation***

**TASK-1**

1. Upon running the bagging model for one set of randomly generated row samples and column samples we got a MSE of  0.09695158102766804 on the training data and OOB score of 11.84766304347826
2. The MSE on the train data is very small as some of the base models would have already seen the train data point as part 
of the randomly chosen sample
3. The MSE obtained from OOB data is also low.

**TASK-2**

1. The house price predicted for the given xq is 19.1 . 

**TASK-3**

1. Upon running 35 iterations of the steps mentioned in Task -1 we come up with a list of 35 values of MSE over train data and 35 values of MSE over OOB data
2. We see that among 35 samples, 95% of them have MSE over train data lying between [ 0.08057169593983873 , 0.13338769798516398 ]
3. We see that among 35 samples, 95% of them have MSE over OOB data lying between [ 8.786385308361155 , 17.761059773656793 ]